### PR TITLE
remove features and related code which were deprecated in dashing

### PR DESCRIPTION
**Pull request by wjwwood**
> 21 additions and 1,069 deletions.

🎉

### CODE REVIEW

File: rclcpp/include/rclcpp/create_subscription.hpp [hunk 1]
```diff
@@ -83,8 +40,10 @@ template<
   typename CallbackT,
   typename AllocatorT = std::allocator<void>,
   typename CallbackMessageT =
-  typename rclcpp::subscription_traits::has_message_type<CallbackT>::type,
+    typename rclcpp::subscription_traits::has_message_type<CallbackT>::type,
   typename SubscriptionT = rclcpp::Subscription<CallbackMessageT, AllocatorT>,
+  typename MessageMemoryStrategyT = rclcpp::message_memory_strategy::MessageMemoryStrategy<
+    CallbackMessageT, AllocatorT>,
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/create_subscription.hpp]: Seems unrelated to the PR title

File: rclcpp/include/rclcpp/create_subscription.hpp [hunk 2]
```diff
@@ -95,9 +54,9 @@ create_subscription(
   const rclcpp::SubscriptionOptionsWithAllocator<AllocatorT> & options = (
     rclcpp::SubscriptionOptionsWithAllocator<AllocatorT>()
   ),
-  typename rclcpp::message_memory_strategy::MessageMemoryStrategy<
-    CallbackMessageT, AllocatorT>::SharedPtr
-  msg_mem_strat = nullptr)
+  typename MessageMemoryStrategyT::SharedPtr msg_mem_strat = (
+    MessageMemoryStrategyT::create_default()
+  ))
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/create_subscription.hpp]: This too

File: rclcpp/include/rclcpp/node.hpp [hunk 3]
```diff
@@ -230,101 +192,25 @@ class Node : public std::enable_shared_from_this<Node>
...
     typename MessageT,
     typename CallbackT,
     typename AllocatorT = std::allocator<void>,
-    typename SubscriptionT = rclcpp::Subscription<
-      typename rclcpp::subscription_traits::has_message_type<CallbackT>::type, AllocatorT>>
+    typename CallbackMessageT =
+      typename rclcpp::subscription_traits::has_message_type<CallbackT>::type,
+    typename SubscriptionT = rclcpp::Subscription<CallbackMessageT, AllocatorT>,
+    typename MessageMemoryStrategyT =
+      rclcpp::message_memory_strategy::MessageMemoryStrategy<CallbackMessageT, AllocatorT>>
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node.hpp]: this

File: rclcpp/include/rclcpp/node.hpp [hunk 4]
```diff
@@ -230,101 +192,25 @@ class Node : public std::enable_shared_from_this<Node>
...
-    size_t qos_history_depth,
-    rclcpp::callback_group::CallbackGroup::SharedPtr group = nullptr,
-    bool ignore_local_publications = false,
-    typename rclcpp::message_memory_strategy::MessageMemoryStrategy<
-      typename rclcpp::subscription_traits::has_message_type<CallbackT>::type, Alloc>::SharedPtr
-    msg_mem_strat = nullptr,
-    std::shared_ptr<Alloc> allocator = nullptr);
+    typename MessageMemoryStrategyT::SharedPtr msg_mem_strat = (
+      MessageMemoryStrategyT::create_default()
+    ));
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node.hpp]: and this

File: rclcpp/include/rclcpp/node_impl.hpp [hunk 5]
```diff
@@ -79,48 +79,20 @@ Node::create_publisher(
...
-}
-
 template<
   typename MessageT,
   typename CallbackT,
   typename AllocatorT,
-  typename SubscriptionT>
+  typename CallbackMessageT,
+  typename SubscriptionT,
+  typename MessageMemoryStrategyT>
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node_impl.hpp]: unrelated

File: rclcpp/include/rclcpp/node_impl.hpp [hunk 6]
```diff
@@ -79,48 +79,20 @@ Node::create_publisher(
...
 std::shared_ptr<SubscriptionT>
 Node::create_subscription(
   const std::string & topic_name,
   const rclcpp::QoS & qos,
   CallbackT && callback,
   const SubscriptionOptionsWithAllocator<AllocatorT> & options,
-  typename rclcpp::message_memory_strategy::MessageMemoryStrategy<
-    typename rclcpp::subscription_traits::has_message_type<CallbackT>::type, AllocatorT>::SharedPtr
-  msg_mem_strat)
+  typename MessageMemoryStrategyT::SharedPtr msg_mem_strat)
```
Comment (ivanpauno) [inline rclcpp/include/rclcpp/node_impl.hpp]: unrelated

**Comment by wjwwood** [comment]
I removed the unrelated commits, I'll do them in a separate pr with some other cleanup commits I want to make. CI (fastrtps only):

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8144/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4130/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6633/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8025/)

**Comment by wjwwood** [comment]
Looks like I need to remove more tests from `test_rclcpp`. Working on that now.

**Comment by wjwwood** [comment]
New CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/8148/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/4134/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/6637/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/8029/)

**Comment by wjwwood** [comment]
All Windows test failures are also on last night's nightly:

https://ci.ros2.org/view/nightly/job/nightly_win_rel/1313/testReport/

**Comment by mkhansen-intel** [comment]
Sorry did I miss a memo / discourse post about this? Removing these broke the Nav2 build.

**Comment by ivanpauno** [comment]
> Sorry did I miss a memo / discourse post about this? Removing these broke the Nav2 build.

I suppose you were getting a deprecated compile warning.
In that case, I don't think any extra announcement was needed.

**Comment by mkhansen-intel** [comment]
Sorry I guess we had disabled the warnings for some reason, I didn't know. In the future, when these changes are made, can it be communicated like "On X date this will be deprecated", so we have some time to pre-emptively fix? I'm not blaming, just asking for some way of communicating. We're trying to keep our master current to the ros2.repos master on a daily basis.

**Comment by jacobperron** [comment]
@mkhansen-intel Not sure if it affects you, but on a related note: https://github.com/ros2/rclpy/pull/431

I agree with @ivanpauno that the warnings should be sufficient notice.
As a general rule, you can expect deprecated API to be removed following the latest major release. E.g. API that is deprecated between now and the release of Eloquent will be removed sometime following the release. I recommend configuring your CI builds to fail on warnings, or communicate them in the results, as a way of being notified.

In general the master branches are volatile by nature, so I think it's difficult to predict/communicate when and what API changes will occur in advanced.

**Comment by wjwwood** [comment]
We also noted that they were deprecated in the release notes of Dashing:

https://index.ros.org/doc/ros2/Releases/Release-Dashing-Diademata/#changes-to-creating-publishers-and-subscriptions

If you have an idea for how we can communicate it better I'm willing to try.

Also, to reiterate what @jacobperron said, master is going to be unstable. We do make sure it compiles with all of our code, but that doesn't mean we won't need to break API's occasionally, for example when tick-tock with deprecation isn't possible or not worth the effort for uncommonly used API's. One day we'll get to 100% API stability on the public API's, but we're not there yet.

It might be a good idea for you guys to build on a snapshot of master instead, and then periodically update and look for issues. On the other hand, it is helpful to know when we break you guys because it might be an accident.
